### ISSUE TITLE
Missing typical optional/nullable operator

### ISSUE THREAD

**Issue by DimaVartanian** [Anviking/Decodable#31]
Hi! Nice library. We're using Argo now and going to try this one out as a possible replacement.

One problem I'm noticing though, is that there doesn't seem to be support for regular nullable functionality. You have 2 nullable (?) operators but neither do what I would expect.

The only nullable operator (for setting an optional var) one would typically need would result in 3 cases:
1. return the object if decode succeeds
2. return nil if the key does not exist
3. return an error if the key exists but decoding fails (ie the JSON object is corrupt/unexpected)

This project does not satisfy case 3 and instead returns nil as well instead of an error(at least according to the documentation). 

In addition, there is the array '?' operator which weeds out objects that failed to decode but in general I think it should follow the same pattern as above (with no partial successes). Allowing a partial success should probably be a special case with its own operator if at all, because in general this would be the result of developer error (unless you are depending on an unstable API).

tl:dr Should this library provide a way to decode an array or custom object in a failable/throwable way which satisfies the 3 above conditions?

Just some thoughts, what do you think?

**Comment by Anviking** [comment]
I would think of case 2 as the dissatisfying constraint. The regular operator (`=>`) has overloads for returning `T?` or `[T]?`, where `nil` is an accepted value. (Or rather it catches `TypeMismatch` Errors where the type is  `NSNull`.) 

Is it actually an accepted practice to not include a key instead of passing nil?

I do think that the array-partial-success-operator `=>?` (btw, thanks for coming up with that description!!!) overload is useful. It would be the result of developer error, but if that error is a very situational one, and is unnoticed and shipped, it could potentially break the entire application because of one unexpected object. Though one could argue that this problem mostly apply to the decoding of the root JSON object where one might as well call `[Foo].decode(json, ignoreInvalidObjects: true)`, not inside the decode-function of another object. 

Regarding the pattern – there is only one other overload – which was the last one to be added, so there is really not an established pattern. (As I argue [here](https://github.com/Anviking/Decodable/blob/master/Decodable/Operators.swift#L52L53)) Though I fully agree on it not being optimal.

Perhaps the `try?` overload of `=>?` could be removed now that we have `try?`, to discourage the usage of it.

It is also possible that the partial-success-array overload could be removed in favor of using flatMap or something instead (imagine `j => "key".flatMap {$0} as [Foo]`). But don't know if that will work.

Okay, I'm now almost convinced that the partial-success-array overload should be removed. And possibly the other one too, removing `=>?`-operator entirely. I will have to play around and think about it when I have time.

Thank you very much for you thoughts!

**Comment by leftspin** [comment]
Hi Anviking, not including a key means "the value of this key has not changed" whereas a key with a null value means "the value of this key has been set to null".

**Comment by Anviking** [comment]
But then a lot more client-side logic would be required? If there would be an overload that meets the criteria (1-3), the unchanged value and the null value would both be represented as nil in your object. By catching an MissingKey error in the current implementation you can differentiate them.

**Comment by Anviking** [comment]
Removed  `=>?` in d8d86cd5dd48a43c44c18e1d4baecbeed34aa4ac.
